returnn/tf/layers/basic.py

```diff
@@ -2695,22 +2693,41 @@ def get_out_data_from_opts(cls, name, sources, axis="T", **kwargs):
     # Just same format.
     return get_concat_sources_data_template(sources, name="%s_output" % name)
 
+  # noinspection PyMethodOverriding
   @classmethod
-  def get_rec_initial_extra_outputs(cls, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
+  def get_rec_initial_extra_outputs(cls, network, batch_dim, rec_layer, axis="T", sources=(), **kwargs):
     """
+    :param returnn.tf.network.TFNetwork network:
     :param tf.Tensor batch_dim:
     :param returnn.tf.layers.rec.RecLayer|LayerBase rec_layer:
     :param str axis:
     :param list[LayerBase] sources:
     :rtype: dict[str,tf.Tensor]
     """
     data = get_concat_sources_data_template(sources)
-    if axis == "T" and data.time_dim_axis is None:
-      # Assume inside RecLayer.
+    if cls._is_recurrent(axis=axis, network=network, data=data):
       assert all(data.shape)
       return {"state": tf.zeros(data.get_batch_shape(batch_dim=batch_dim), dtype=data.dtype)}
     return {}
 
+  @classmethod
+  def _is_recurrent(cls, axis, network, data):
+    """
+    :param str|DimensionTag axis:
+    :param returnn.tf.network.TFNetwork network:
+    :param Data data:
+    :rtype: bool
+    """
+    if not network.is_inside_rec_layer(inside_loop=False):
+      return False
+    if axis.lower() == "t" and data.time_dim_axis is None:  # legacy fallback
+      return True
+    if axis.lower() == ":i":
```

Comment:
I'm not sure that we really need this. I remember this was some suggestion in the self-attention generalization discussion but we actually did not implement anything like that.

And when you just use dim tags consistently (e.g. specify `axis` of the outer `RecLayer`), you don't need this special string logic here, you can directly pass the dim tag of the outer rec layer.

Reply:
Okay yeah, I'll just use the dim tags and remove the `:i` case.